src/mlpack/methods/ann/loss_functions/mean_absolute_error_impl.hpp

```diff
@@ -0,0 +1,66 @@
+/**
+ * @file methods/ann/loss_functions/mean_absolute_error_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of the Mean Absolute Error function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_ERROR_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_MEAN_ABSOLUTE_ERROR_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "mean_absolute_error.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MeanAbsoluteError<InputDataType, OutputDataType>::MeanAbsoluteError(const bool reduction):
+  reduction(reduction)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+typename InputType::elem_type
+MeanAbsoluteError<InputDataType, OutputDataType>::Forward(
+    const InputType& input,
+    const TargetType& target)
+{
+  if (reduction)
+    return arma::accu(arma::abs((input - target) / input));
```

Comment:
@iamshnoo is the actual implementation of the error function right ?

Reply:
If I understand correctly, you probably need to divide by the number of elements in the input, instead of directly dividing by the input matrix itself.

Reply:
[here](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error) it is mentioned that it is divided by the input, for the common MAE function but in this [paper](https://arxiv.org/pdf/1605.02541.pdf) it is mentioned that it is divided by the target.

Reply:
You could see the later part on page 24 of the paper.

Reply:
In the Wikipedia article, the actual value is the ```target``` which matches the one from the paper. So, you should be dividing by ```target```, instead of ```input```.
Also, the name of the error is Mean Absolute Percentage Error, and not Mean Absolute Error (as you have used everywhere), so, it would be less confusing if you renamed the class correctly.

Reply:
Sure

Reply:
You can add the citation of the paper, yes. 
For an example of how to do so, you can take a look at ```mlpack/src/mlpack/methods/ann/layer/batch_norm.hpp```.

Reply:
okay great.

Reply:
What is the 0.1 in the CheckMatrices [here](https://github.com/mlpack/mlpack/pull/2495/commits/79a1fc588fa20658cadad99d170b72423fb830fd#diff-1f48422898330115de61fe91a3cc4242R728)

Reply:
Hey @iamshnoo i have pushed the final commit from my side, if you are free can you check it ?